File: example_code/topic2/gauss.py
# estimate Gaussian integral using uniform and importance sampling

from math import pi, sqrt
from random import random, normalvariate

double f_over_w(x):
    return x**2

# get input parameters from user
N = int(input(" Enter number of points N: "))
L = float(input(" Enter cut-off L for uniform sampling: "))

# uniform sampling
avg = 0.0
var = 0.0
for i in range(N):
    x = (2 * random() - 1) * L;
    fx = f(x)
    avg += fx
    var += fx**2

avg /= N
var /= N
var = var - avg**2
print()
print("   Uniform sampling: " + str(2 * L * avg) + "  +-  "
    + str(2 * L * sqrt(var / N)))

# importance sampling
avg = var = 0.0
for i in range(N):
    x = normalvariate();
    fx = f_over_w(x)
    avg += fx
    var += fx**2

avg /= N
var /= N
var = var - avg**2
print("Importance sampling: " + str(avg) + "  +-  " + str(sqrt(var / N)))
print("       Exact answer: " + str(1.0))

File: numerical_analysis/example_code/topic2/gauss.cpp
// estimate Gaussian integral using uniform and importance sampling

#include <cmath>
#include <cstdlib>
#include <iostream>
using namespace std;

const double pi = 4 * atan(1.0);

#include "../tools/random.hpp"

double f(double x)
{
    return x * x * exp(- x * x / 2) / sqrt(2 * pi);
}

double f_over_w(double x)
{
    return x * x;
}

int main() {

    // get input parameters from user
    cout << "Enter number of points N: ";
    int N;
    cin >> N;
    cout << "Enter cut-off L for uniform sampling: ";
    double L;
    cin >> L;

    // uniform sampling
    double avg = 0;
    double var = 0;
    for (int i = 0; i < N; i++) {
        double x = (2 * uniform_dist() - 1) * L;
        double fx = f(x);
        avg += fx;
        var += fx * fx;
    }
    avg /= N;
    var /= N;
    var = var - avg * avg;
    cout << "\n   Uniform sampling: " << 2 * L * avg << "  +-  "
         << 2 * L * sqrt(var / N) << endl;

    // importance sampling
    avg = var = 0;
    for (int i = 0; i < N; i++) {
        double x = normal_dist();
        double fx = f_over_w(x);
        avg += fx;
        var += fx * fx;
    }
    avg /= N;
    var /= N;
    var = var - avg * avg;
    cout << "Importance sampling: " << avg << "  +-  "
         << sqrt(var / N) << endl;

    cout << "       Exact answer: " << 1.0 << endl;
}
